src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -106,13 +105,12 @@ protected override void OnInit()
         // called by BufferedFileStreamStrategy
         internal override void OnBufferAllocated(byte[] buffer)
         {
-            Debug.Assert(_preallocatedOverlapped == null);
+            Debug.Assert(_reusableValueTaskSource == null);
 
-            _preallocatedOverlapped = new PreAllocatedOverlapped(ValueTaskSource.s_ioCallback, this, buffer);
+            _reusableValueTaskSource = new ValueTaskSource(this, buffer);
         }
```

Comment:
If memory serves, this OnBufferAllocated is one of the things that prevented us from just using BufferedStream.

We also have complexity/cost in the implementation trying to account for differences between preallocated overlappeds and whether they wrap the internal buffer.

Thought:
What if we create the preallocated overlappeds never wrapping a buffer and always using MemoryHandle.Pin.  We then change BufferedFileStreamStrategy to either a) allocate from the pinned-object heap or b) create a buffer that it itself pins.  It then creates a Memory using MemoryMarshal.CreateFromPinnedArray, and passes that through to this strategy, such that calls to Pin become nops.

In theory that should simplify the code, improve separation of concerns, and potentially help with perf as well.

Reply:
I like this idea. I have implemented it and all tests are passing, will have benchmark results in 20 minutes from now